src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerAssemblyLoader.Core.cs

```diff
@@ -83,25 +87,45 @@ private partial bool IsMatch(AssemblyName requestedName, AssemblyName candidateN
 
         internal DirectoryLoadContext[] GetDirectoryLoadContextsSnapshot()
         {
+            CheckIfDisposed();
+
             lock (_guard)
             {
                 return _loadContextByDirectory.Values.OrderBy(v => v.Directory).ToArray();
             }
         }
 
-        internal void UnloadAll()
+        partial void DisposeWorker()
         {
-            List<DirectoryLoadContext> contexts;
+            var contexts = ArrayBuilder<DirectoryLoadContext>.GetInstance();
             lock (_guard)
             {
-                contexts = _loadContextByDirectory.Values.ToList();
+                foreach (var (_, context) in _loadContextByDirectory)
+                    contexts.Add(context);
+
                 _loadContextByDirectory.Clear();
             }
 
             foreach (var context in contexts)
             {
+#if WORKSPACE
```

Comment:
If `ReportAndCatch` is available in the compiler layer I would just remove the `#if` here. It's fine for us to NFW.

Reply:
done.